src/v/kafka/server/handlers/fetch.cc

```diff
@@ -200,23 +200,15 @@ static ss::future<read_result> do_read_from_ntp(
           error_code::unknown_topic_or_partition);
     }
 
-    auto high_watermark = kafka_partition->high_watermark();
-
-    auto max_offset = high_watermark < model::offset(0) ? model::offset(0)
-                                                        : high_watermark;
-
     if (config::shard_local_cfg().enable_transactions.value()) {
         if (
           ntp_config.cfg.isolation_level
           == model::isolation_level::read_committed) {
             ntp_config.cfg.max_offset = kafka_partition->last_stable_offset();
-            max_offset = kafka_partition->last_stable_offset();
```

Comment:
It's a good catch, I've checked what Kafka does instead of returning out_of_range when start offset is ahead of LSO and it [returns empty batch](https://github.com/apache/kafka/blob/3944f9664783e0cfbf5a00d49ac91192404df35b/core/src/main/scala/kafka/log/Log.scala#L1210).

One thing that concerns me is that we used to return `offset_out_of_range` when `ntp_config.cfg.start_offset` > `ntp_config.cfg.max_offset` and now we pass it down to `read_from_partition`. Have you checked that the lower layers handle this correctly (start_offset > max_offset) and returns empty batch?

Reply:
exactly, we do return empty reader when there are nothing to read